### PR TITLE
feat: mask values typed into input[type=password] in a reporter

### LINKED ISSUE #6014: Mask values typed into masked inputs when sending action logs to a reporter
### What is your Test Scenario?
I type my password with `t.typeText()` and use a reporter that stores action logs.

### What are you suggesting?
I'd like sensitive data to be automatically masked. We could look at the input where you type. If it's masked, hide the value from the logs.

We may also want to introduce an option that overrides this behavior and allows for logging individual/all? sensitive values.

### PR DESCRIPTION
**Pull request by wentwrong**
## Purpose
Values typed into input[type=password] should be replaced by a placeholder before sending action logs with these values to a reporter. Also the `pressKey` and the `typeText` actions have the `confidential` flag. If it set to 'true', then any typed text & pressed keys would be replaced by a placeholder in action logs.

## Approach
Detect what is the target element (either specified in the `typeText` action or current active element for the `pressKey` action). If it's the input which type is 'password' or the `confidential` flag is set to true, then replace info with a placeholder.

## References
Closes #6014

## Pre-Merge TODO
- [x] Write tests for your proposed changes
- [ ] Make sure that existing tests do not fail

### CODE REVIEW

File: src/utils/is-password-input.ts [hunk 1]
```diff
@@ -0,0 +1,3 @@
+export default function isPasswordInput (node: any): boolean {
```
Comment (miherlosev) [inline src/utils/is-password-input.ts:1]: `(node: any)`
Can we use a strict type here?

**Review by miherlosev — CHANGES_REQUESTED** [review]

File: src/test-run/commands/options.js [hunk 2]
```diff
@@ -254,3 +256,20 @@ export class AssertionOptions extends Assignable {
         ];
     }
 }
+
+// Press
+export class PressOptions extends ActionOptions {
+    constructor (obj, validate) {
+        super();
+
+        this.confidential = void 0;
```
[the remark this answers is not in the document]
Comment (wentwrong) [reply src/test-run/commands/options.js:265]: `void 0` is necessary as a default value. It's needed to indicate that a user didn't set this flag explicitly. 

I've tried to use `enum` for this flag, but it looks overcomplicated in my opinion (ts-hack to create 3-state boolean):
```ts
enum ConfidentialSetting {
    False,
    True,
    CheckRequired
}
```
Now by default, this option will be shown in the "options" property of the command: 
```js
{
    type:    TYPE.typeText,
    options: {
        confidential: 2 // or the "CheckRequired" string (not better)
    }
}
```
Whereas `void 0` wouldn't be set at all. I think it's better to stick with the `void 0` variant.

File: src/client/driver/driver.js [hunk 3]
```diff
@@ -268,6 +273,13 @@ export default class Driver extends serviceUtils.EventEmitter {
         return Promise.resolve();
     }
 
+    async _getActiveElement () {
+        const activeElement = domUtils.getActiveElement();
+        const encodedResult = this.replicator.encode(activeElement);
```
Comment (miherlosev) [inline src/client/driver/driver.js:278]: `return this.replicator.encode(activeElement);`

File: src/test-run/commands/options.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (miherlosev) [reply src/test-run/commands/options.js:265]: It's not clear what the `void 0` value means in the context of the `confidential` option.
So, in my opinion, the `enum ConfidentialSetting` is a good variant.

@AndreyBelym what do you think about it?

File: src/tsconfig.json [hunk 4]
```diff
@@ -1,5 +1,9 @@
 {
-  "include": [ "../@types", "./"],
+  "include": [
+    "../@types",
+    "../ts-defs-src",
```
Comment (miherlosev) [inline src/tsconfig.json:4]: `"../ts-defs-src"`
Why does it necessary? I as remember we have a separate Gulp [step](https://github.com/DevExpress/testcafe/blob/master/Gulpfile.js#L227) for the TypeScript definitions.

**Review by miherlosev — CHANGES_REQUESTED** [review]

File: src/test-run/commands/options.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (AndreyBelym) [reply src/test-run/commands/options.js:265]: IMHO:
TLDR; I'm against a tri-state enum:

1. In this enum, 2 of 3 values just replicate boolean options. For me, it's a sign of redundant complexity. 

2. I can't see any case when I would want to specify `ConfidentialSetting.CheckRequired` manually. Why should we give it a name, if a user won't interact with it?

3. I can describe easily the boolean configuration:
   If `true`, we hide the input value
   If `false`, we show the input value
   By default, it's `true` when the target selector is input[type=password], `false` otherwise.

**Review by AndreyBelym — APPROVED** [review]

File: src/tsconfig.json (hunk 4 above)
Comment (AndreyBelym) [reply src/tsconfig.json:4]: I guess it should test the compatibility between our end-user TS defs and our source-level defs.

File: src/test-run/commands/options.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (AndreyBelym) [reply src/test-run/commands/options.js:265]: Well, about point 2: I see a case, if we implement some global-level options, you may want to reset this flag to default in some cases. But since `undefined` is a common way to reset something to default in JS everywhere, I don't see it as a problem.

[the remark this answers is not in the document]
Comment (miherlosev) [reply src/test-run/commands/options.js:265]: Ok. Let it be.

File: src/tsconfig.json (hunk 4 above)
Comment (wentwrong) [reply src/tsconfig.json:4]: I've added the end user TS defs in order to reuse them in the source code and don't maintain two versions of the same TS defs.

For example  [`NodeSnapshot`](https://github.com/DevExpress/testcafe/pull/6191/files#diff-997e4f44307a245049ef83b59a672b489463b06bd4a1d571557cbb3bed03845fR1) was already defined in the end-user TS defs: 
https://github.com/DevExpress/testcafe/blob/c72b8cd2ad88f09d8194a5f48ee3d954fb2b0aae/ts-defs-src/test-api/node-snapshot.d.ts#L29 
Also the [`CompilerOptions`](https://github.com/DevExpress/testcafe/pull/6191/files#diff-0269d3bc8b9fe8b07e269b1f3928a88d59eaf0fe4f52d6a63ee3d9d7dc5f302fL15) interface was defined too: https://github.com/DevExpress/testcafe/blob/c72b8cd2ad88f09d8194a5f48ee3d954fb2b0aae/ts-defs-src/runner-api/runner-api.d.ts#L108

**Review by miherlosev — CHANGES_REQUESTED** [review]
The last issue - https://github.com/DevExpress/testcafe/pull/6191/files#r628129446

**Review by miherlosev — APPROVED** [review]
